File: engine/src/main/java/good/damn/engine/opengl/MGObject3d.java
package good.damn.engine.opengl;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.nio.FloatBuffer;
import java.nio.IntBuffer;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

import good.damn.engine.utils.MGUtilsBuffer;
import good.damn.engine.utils.MGUtilsFile;

public final class MGObject3d {

    private static final Charset PATH_CHARSET = StandardCharsets.UTF_8;

    @NonNull
    public final FloatBuffer vertices;

    @NonNull
    public final IntBuffer indices;

    @Nullable
    public final String[] texturesDiffuseFileName;

    @Nullable
    public final String[] texturesMetallicFileName;

    @Nullable
    public final String[] texturesEmissiveFileName;

    public MGObject3d(
        @NonNull final FloatBuffer vertices,
        @NonNull final int[] indices,
        @Nullable final String[] texturesDiffuseFileName,
        @Nullable final String[] texturesMetallicFileName,
        @Nullable final String[] texturesEmissiveFileName
    ) {
        this.vertices = vertices;
        this.indices = MGUtilsBuffer.Companion.createInt(
            indices
        );

        this.texturesDiffuseFileName = texturesDiffuseFileName;
        this.texturesMetallicFileName = texturesMetallicFileName;
        this.texturesEmissiveFileName = texturesEmissiveFileName;
    }

    public MGObject3d(
        @NonNull final float[] vertices,
        @NonNull final int[] indices,
        @Nullable final String[] texturesDiffuseFileName,
        @Nullable final String[] texturesMetallicFileName,
        @Nullable final String[] texturesEmissiveFileName
    ) {
        this.vertices = MGUtilsBuffer.Companion.createFloat(
            vertices
        );

        this.indices = MGUtilsBuffer.Companion.createInt(
            indices
        );

        this.texturesDiffuseFileName = texturesDiffuseFileName;
        this.texturesMetallicFileName = texturesMetallicFileName;
        this.texturesEmissiveFileName = texturesEmissiveFileName;
    }

    static {
        System.loadLibrary(
            "engine"
        );
    }

    @Nullable
    public static MGObject3d[] createFromAssets(
        @NonNull final String localPath
    ) throws Exception {
        @NonNull final File filePub = MGUtilsFile.Companion.getPublicFile(
            localPath
        );

        if (!filePub.exists()) {
            throw new Exception("No such file: " + filePub.getPath());
        }

        return createFromPath(
            filePub.getPath()
        );
    }

    @Nullable
    public static MGObject3d[] createFromPath(
        @NonNull String path
    ) {
        return createFromPath(
            path.getBytes(
                PATH_CHARSET
            )
        );
    }

    @Nullable
    private static native MGObject3d[] createFromPath(
        @NonNull byte[] path
    );
}

File: engine/src/main/java/good/damn/engine/opengl/MGObject3D.kt
package good.damn.engine.opengl

import good.damn.engine.MGEngine
import good.damn.engine.utils.MGUtilsBuffer
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.lang.Exception
import java.nio.FloatBuffer
import java.nio.IntBuffer
import java.util.Vector

class MGObject3D(
    val vertices: FloatBuffer,
    val indices: IntBuffer
) {
    companion object {

        fun createFromAssets(
            path: String
        ): MGObject3D {
            return createFromStream(
                MGEngine.ASSETS.open(
                    path
                )
            )
        }

        fun createFromStream(
            inp: InputStream
        ): MGObject3D {
            val vertices: Vector<Float> = Vector()
            val normals: Vector<Float> = Vector()
            val textures: Vector<Float> = Vector()
            val faces: Vector<String> = Vector()

            var reader: BufferedReader? = null

            try {
                val inStream = InputStreamReader(
                    inp
                )
                reader = BufferedReader(inStream)

                // read file until EOF
                var line = reader.readLine()
                while (line != null) {
                    val parts = line.split("\\s+".toRegex())
                    when (parts[0]) {
                        "v" -> {
                            // vertices
                            vertices.add(parts[1].toFloat())
                            vertices.add(parts[2].toFloat())
                            vertices.add(parts[3].toFloat())
                        }
                        "vt" -> {
                            // textures
                            textures.add(parts[1].toFloat())
                            textures.add(parts[2].toFloat())
                        }
                        "vn" -> {
                            // normals
                            normals.add(parts[1].toFloat())
                            normals.add(parts[2].toFloat())
                            normals.add(parts[3].toFloat())
                        }
                        "f" -> {
                            // faces: vertex/texture/normal
                            faces.add(parts[1])
                            faces.add(parts[2])
                            faces.add(parts[3])
                        }
                    }
                    line = reader.readLine()
                }
            } catch (e: Exception) {
                // cannot load or read file
            } finally {
                if (reader != null) {
                    try {
                        reader.close()
                    } catch (e: Exception) {
                        //log the exception
                    }
                }
            }

            // 3 - position
            // 2 - texture coords
            // 3 - normals
            val bufferVertices = MGUtilsBuffer.allocateFloat(
                faces.size * 8
            )
            val bufferIndices = MGUtilsBuffer.allocateInt(
                faces.size
            )

            var posIndex = 0

            for ((index, face) in faces.withIndex()) {
                val parts = face.split("/")
                val vertexIndex = (parts[0].toShort() - 1).toShort()
                bufferIndices.put(
                    index,
                    index
                )

                var i = 3 * vertexIndex

                bufferVertices.put(
                    posIndex++,
                    vertices[i++]
                )

                bufferVertices.put(
                    posIndex++,
                    vertices[i++]
                )

                bufferVertices.put(
                    posIndex++,
                    vertices[i]
                )

                i = 2 * (parts[1].toInt() - 1)
                bufferVertices.put(
                    posIndex++,
                    textures[i++]
                )
                bufferVertices.put(
                    posIndex++,
                    1f - textures[i]
                )

                i = 3 * (parts[2].toInt() - 1)
                bufferVertices.put(
                    posIndex++,
                    normals[i++]
                )
                bufferVertices.put(
                    posIndex++,
                    normals[i++]
                )
                bufferVertices.put(
                    posIndex++,
                    normals[i]
                )
            }

            bufferVertices.position(0)
            bufferIndices.position(0)

            return MGObject3D(
                bufferVertices,
                bufferIndices
            )
        }

    }
}